File: src/server/orchestrator/limitMonitor.ts
import { containsLimitSignal, stripAnsi } from './resumeDetector.js';
import { sumJsonlSizes } from './mountActivity.js';

export interface LimitMonitorOptions {
  /** Dir to watch for Claude Code's session JSONL writes (activity signal). */
  mountDir: string;
  /** Rolling ANSI-stripped buffer size, in bytes. */
  logBufferBytes?: number;
  /** Required idle time on mount dir before firing. */
  idleMs?: number;
  /** Absolute warmup after `start()` before firing is allowed. */
  warmupMs?: number;
  /** Check cadence. */
  checkMs?: number;
  /** Fired exactly once. */
  onDetect: () => void;
  /** Optional clock injection for tests. */
  now?: () => number;
}

/**
 * Watches the container's TTY stream for a rate-limit message and fires once
 * when Claude is clearly stuck on that message.
 *
 * The detector uses two independent signals to avoid false positives:
 *  1. The rolling ANSI-stripped tail contains a limit phrase.
 *  2. Claude has been idle — no new bytes appended to any session JSONL in
 *     `mountDir` for `idleMs` — AND we're past an absolute `warmupMs` from
 *     start, which is long enough that any prompt echo has scrolled out of
 *     the rolling window on a live run.
 *
 * Both must hold simultaneously. This rules out false positives from user
 * prompts that literally contain "hit your limit …" (they're echoed early,
 * but Claude is making API calls so the mount dir isn't idle).
 */
export class LimitMonitor {
  private buf = '';
  private fired = false;
  private startedAt = 0;
  private lastActivityAt = 0;
  private lastTotalSize = 0;
  private timer: NodeJS.Timeout | null = null;
  private readonly logBufferBytes: number;
  private readonly idleMs: number;
  private readonly warmupMs: number;
  private readonly checkMs: number;
  private readonly now: () => number;

  constructor(private opts: LimitMonitorOptions) {
    this.logBufferBytes = opts.logBufferBytes ?? 16 * 1024;
    this.idleMs = opts.idleMs ?? 15_000;
    this.warmupMs = opts.warmupMs ?? 60_000;
    this.checkMs = opts.checkMs ?? 3_000;
    this.now = opts.now ?? Date.now;
  }

  start(): void {
    if (this.timer) return;
    this.startedAt = this.now();
    this.lastActivityAt = this.startedAt;
    this.lastTotalSize = this.currentMountSize();
    const tick = () => {
      if (!this.timer) return;
      try { this.check(); } catch { /* swallow — best-effort */ }
      this.timer = setTimeout(tick, this.checkMs);
    };
    this.timer = setTimeout(tick, this.checkMs);
  }

  stop(): void {
    if (this.timer) clearTimeout(this.timer);
    this.timer = null;
  }

  /** Feed a chunk of raw TTY bytes (ANSI allowed). Called per attach 'data'. */
  feedLog(chunk: Uint8Array): void {
    if (this.fired) return;
    const text = Buffer.from(chunk).toString('utf8');
    const stripped = stripAnsi(text);
    if (stripped.length === 0) return;
    this.buf = (this.buf + stripped).slice(-this.logBufferBytes);
  }

  /** Force a check now — used by tests. */
  checkNow(): void { this.check(); }

  private check(): void {
    if (this.fired) return;
    const now = this.now();
    // Update activity signal: has any session file grown since last tick?
    const size = this.currentMountSize();
    if (size > this.lastTotalSize) {
      this.lastActivityAt = now;
      this.lastTotalSize = size;
    }
    if (now - this.startedAt < this.warmupMs) return;
    if (now - this.lastActivityAt < this.idleMs) return;
    if (!containsLimitSignal(this.buf)) return;
    this.fired = true;
    this.stop();
    this.opts.onDetect();
  }

  private currentMountSize(): number {
    return sumJsonlSizes(this.opts.mountDir);
  }
}

File: server-elixir/lib/fbi/orchestrator/limit_monitor.ex
defmodule FBI.Orchestrator.LimitMonitor do
  @moduledoc "Port of src/server/orchestrator/limitMonitor.ts."
  use GenServer

  @log_buffer_bytes 16 * 1024
  @idle_ms 15_000
  @warmup_ms 60_000
  @check_ms 3_000

  defstruct [
    :mount_dir,
    :on_detect,
    :log_buffer_bytes,
    :idle_ms,
    :warmup_ms,
    :check_ms,
    buf: "",
    fired: false,
    started_at: 0,
    last_activity_at: 0,
    last_total_size: 0,
    timer: nil
  ]

  def start_link(opts) do
    GenServer.start_link(__MODULE__, opts)
  end

  def stop(pid), do: GenServer.stop(pid)

  def feed_log(pid, chunk) when is_pid(pid) do
    GenServer.cast(pid, {:feed_log, chunk})
  end

  @impl true
  def init(opts) do
    now = System.monotonic_time(:millisecond)
    mount_dir = Keyword.fetch!(opts, :mount_dir)

    state = %__MODULE__{
      mount_dir: mount_dir,
      on_detect: Keyword.fetch!(opts, :on_detect),
      log_buffer_bytes: Keyword.get(opts, :log_buffer_bytes, @log_buffer_bytes),
      idle_ms: Keyword.get(opts, :idle_ms, @idle_ms),
      warmup_ms: Keyword.get(opts, :warmup_ms, @warmup_ms),
      check_ms: Keyword.get(opts, :check_ms, @check_ms),
      started_at: now,
      last_activity_at: now,
      last_total_size: mount_size(mount_dir)
    }

    timer = Process.send_after(self(), :tick, state.check_ms)
    {:ok, %{state | timer: timer}}
  end

  @impl true
  def handle_cast({:feed_log, chunk}, %{fired: false} = state) do
    stripped = FBI.Orchestrator.ResumeDetector.strip_ansi(chunk)

    if stripped == "" do
      {:noreply, state}
    else
      new_buf =
        (state.buf <> stripped) |> String.slice(-state.log_buffer_bytes, state.log_buffer_bytes)

      {:noreply, %{state | buf: new_buf}}
    end
  end

  def handle_cast({:feed_log, _}, state), do: {:noreply, state}

  @impl true
  def handle_info(:tick, %{fired: true} = state), do: {:noreply, state}

  def handle_info(:tick, state) do
    now = System.monotonic_time(:millisecond)
    size = mount_size(state.mount_dir)

    state =
      if size > state.last_total_size do
        %{state | last_activity_at: now, last_total_size: size}
      else
        state
      end

    state =
      if now - state.started_at >= state.warmup_ms and
           now - state.last_activity_at >= state.idle_ms and
           FBI.Orchestrator.ResumeDetector.contains_limit_signal(state.buf) do
        state.on_detect.()
        %{state | fired: true}
      else
        state
      end

    timer = if state.fired, do: nil, else: Process.send_after(self(), :tick, state.check_ms)
    {:noreply, %{state | timer: timer}}
  end

  defp mount_size(dir) do
    case File.ls(dir) do
      {:error, _} ->
        0

      {:ok, files} ->
        Enum.reduce(files, 0, fn f, acc ->
          case File.stat(Path.join(dir, f)) do
            {:ok, %{size: s, type: :regular}} -> acc + s
            _ -> acc
          end
        end)
    end
  end
end
